### PR TITLE
JAMES-2726 Add In-memory image creation

### PR DESCRIPTION
**Pull request by blackheaven**

### CODE REVIEW

File: dockerfiles/run/guice/memory/destination/conf/elasticsearch.properties [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+#  KIND, either express or implied.  See the License for the
+#  specific language governing permissions and limitations
+#  under the License.
+
+#  This template file can be used as example for James Server configuration
+#  DO NOT USE IT AS SUCH AND ADAPT IT TO YOUR NEEDS
+
+# Configuration file for ElasticSearch
+
+elasticsearch.masterHost=elasticsearch
```
Comment (chibenwa) [inline dockerfiles/run/guice/memory/destination/conf/elasticsearch.properties]: This file is propbably not needed (except metric part)

File: dockerfiles/run/guice/memory/destination/conf/mailetcontainer.xml [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+            </mailet>
+        </processor>
+
+        <processor state="error" enableJmx="true">
+            <mailet match="All" class="MetricsMailet">
+                <metricName>mailetContainerErrors</metricName>
+            </mailet>
+            <mailet match="All" class="Bounce"/>
+            <mailet match="All" class="ToRepository">
+                <repositoryPath>cassandra://var/mail/error/</repositoryPath>
```
Comment (chibenwa) [inline dockerfiles/run/guice/memory/destination/conf/mailetcontainer.xml]: cassandra???

File: dockerfiles/run/guice/memory/destination/conf/mailetcontainer.xml [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+        <processor state="relay-denied" enableJmx="true">
+            <mailet match="All" class="MetricsMailet">
+                <metricName>mailetContainerRelayDenied</metricName>
+            </mailet>
+            <mailet match="All" class="Bounce">
+                <attachment>none</attachment>
+            </mailet>
+            <mailet match="All" class="ToRepository">
+                <repositoryPath>cassandra://var/mail/relay-denied/</repositoryPath>
```
Comment (chibenwa) [inline dockerfiles/run/guice/memory/destination/conf/mailetcontainer.xml]: idem

File: dockerfiles/run/guice/memory/destination/conf/mailetcontainer.xml [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+                <metricName>bounces</metricName>
+            </mailet>
+            <mailet match="All" class="DSNBounce">
+                <passThrough>false</passThrough>
+            </mailet>
+        </processor>
+
+        <processor state="rrt-error" enableJmx="false">
+            <mailet match="All" class="ToRepository">
+                <repositoryPath>cassandra://var/mail/rrt-error/</repositoryPath>
```
Comment (chibenwa) [inline dockerfiles/run/guice/memory/destination/conf/mailetcontainer.xml]: idem

File: dockerfiles/run/guice/memory/destination/conf/rabbitmq.properties [hunk 5]
```diff
@@ -0,0 +1,2 @@
+uri=amqp://james:james@rabbitmq_host:5672
```
Comment (chibenwa) [inline dockerfiles/run/guice/memory/destination/conf/rabbitmq.properties]: Probably not needed?

File: dockerfiles/run/guice/memory/destination/conf/tika.properties [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+#  software distributed under the License is distributed on an
+#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+#  KIND, either express or implied.  See the License for the
+#  specific language governing permissions and limitations
+#  under the License.
+
+tika.enabled=true
+tika.host=tika
+tika.port=9998
+tika.timeoutInMillis=3000
```
Comment (chibenwa) [inline dockerfiles/run/guice/memory/destination/conf/tika.properties]: Not needed?

File: dockerfiles/compilation/java-8/compile.sh [hunk 7]
```diff
@@ -104,6 +105,14 @@ if [ $? -eq 0 ]; then
       cp -r server/container/cli/target/james-server-cli.lib $JPA_SMTP_DESTINATION || true
    fi
 
+   if [ -d "$MEMORY_DESTINATION" ]; then
+      echo "Copying cassandra JARs"
```
Comment (trantienduchn) [inline dockerfiles/compilation/java-8/compile.sh]: s/cassandra/memory

File: dockerfiles/run/guice/memory/destination/conf/elasticsearch.properties (hunk 1 above)
Comment (trantienduchn) [reply dockerfiles/run/guice/memory/destination/conf/elasticsearch.properties]: can you explain? don't we need elasticsearch with memory product?

File: dockerfiles/run/guice/memory/destination/conf/listeners.xml [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+<listeners>
+  <listener>
+    <class>org.apache.james.mailbox.spamassassin.SpamAssassinListener</class>
+    <async>true</async>
+  </listener>
+  <listener>
+    <class>org.apache.james.mailbox.cassandra.MailboxOperationLoggingListener</class>
+  </listener>
+</listeners>
```
Comment (trantienduchn) [inline dockerfiles/run/guice/memory/destination/conf/listeners.xml]: I think we need a section of predeletion hooks configuration. In here and other conf/listeners.xml in other products.
```
  ...
  </listener>
  <preDeletionHook>
    <class>org.apache.james.vault.DeletedMessageVaultHook</class>
  </preDeletionHook>
```

File: dockerfiles/run/guice/memory/destination/conf/logback.xml [hunk 9]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (trantienduchn) [inline dockerfiles/run/guice/memory/destination/conf/logback.xml]: missing apache license

File: dockerfiles/run/guice/memory/destination/conf/listeners.xml (hunk 8 above)
Comment (chibenwa) [reply dockerfiles/run/guice/memory/destination/conf/listeners.xml]: Can you open a PR for fixing other configurations @trantienduchn ?

Comment (trantienduchn) [reply dockerfiles/run/guice/memory/destination/conf/listeners.xml]: here #2290

File: server/container/guice/memory-guice/src/main/java/org/apache/james/MemoryJamesServerMain.java [hunk 10]
```diff
@@ -115,6 +116,7 @@ public static void main(String[] args) throws Exception {
             .build();
 
         GuiceJamesServer.forConfiguration(configuration)
+            .combineWith(new LuceneSearchMailboxModule())
```
Comment (chibenwa) [inline server/container/guice/memory-guice/src/main/java/org/apache/james/MemoryJamesServerMain.java]: No! It's inMemory not lucene

You by the way are lying to your command line users by reindeing docs in a non used searched index

This is madness.

If you really want a CLI, please bind a ThrowingSearchIndex instead - everybody gonna be healthier from this choice.

Note : maybe CLI and thus JMX is not needed at all. Thoughts @mbaechler ?

File: server/container/guice/memory-guice/src/main/java/org/apache/james/modules/data/MemoryDataModule.java [hunk 11]
```diff
@@ -54,7 +54,7 @@
 
 public class MemoryDataModule extends AbstractModule {
     private static final MailRepositoryStoreConfiguration.Item FILE_MAILREPOSITORY_DEFAULT_DECLARATION = new MailRepositoryStoreConfiguration.Item(
-        ImmutableList.of(new Protocol("file")),
+        ImmutableList.of(new Protocol("file"), new Protocol("memory")),
```
Comment (trantienduchn) [inline server/container/guice/memory-guice/src/main/java/org/apache/james/modules/data/MemoryDataModule.java]: why do you need two protocols ?

Comment (blackheaven) [reply server/container/guice/memory-guice/src/main/java/org/apache/james/modules/data/MemoryDataModule.java]: I hopped that we have a Memory protocol, it seems not

File: server/container/guice/memory-guice/src/main/java/org/apache/james/modules/data/MemoryDataModule.java [hunk 12]
```diff
@@ -54,7 +54,7 @@
...
         FileMailRepository.class.getName(),
```
Comment (chibenwa) [inline server/container/guice/memory-guice/src/main/java/org/apache/james/modules/data/MemoryDataModule.java]: This one could be `MemoryMailRepository` no? (and memory above)

File: dockerfiles/run/guice/memory/destination/conf/mailrepositorystore.xml [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+<mailrepositorystore>
+    <defaultProtocol>file</defaultProtocol>
+    <mailrepositories>
+        <mailrepository class="org.apache.james.mailrepository.file.FileMailRepository">
+            <protocols>
+                <protocol>file</protocol>
+            </protocols>
+        </mailrepository>
+    </mailrepositories>
```
Comment (chibenwa) [inline dockerfiles/run/guice/memory/destination/conf/mailrepositorystore.xml]: Please add

```
    <mailrepositories>
        <mailrepository class="org.apache.james.mailrepository.memory.MemoryMailRepository">
            <protocols>
                <protocol>memory</protocol>
            </protocols>
        </mailrepository>
    </mailrepositories>
```

to gain access to Memory mail repositories

 - and drop me this aweful `FileMailRepository` + file protocol ^^

File: server/container/guice/memory-guice/src/main/java/org/apache/james/modules/data/MemoryDataModule.java [hunk 14]
```diff
@@ -54,8 +54,8 @@
 
 public class MemoryDataModule extends AbstractModule {
     private static final MailRepositoryStoreConfiguration.Item FILE_MAILREPOSITORY_DEFAULT_DECLARATION = new MailRepositoryStoreConfiguration.Item(
```
Comment (chibenwa) [inline server/container/guice/memory-guice/src/main/java/org/apache/james/modules/data/MemoryDataModule.java]: It's now `MEMORY_MAILREPOSITORY_DEFAULT_DECLARATION`

**Pull request merged**